test/testutil.c

```diff
@@ -365,21 +365,31 @@ int test_str_ne(const char *file, int line, const char *st1, const char *st2,
 
 /*
  * We could use OPENSSL_buf2hexstr() to do this but trying to allocate memory
- * in a failure state isn't generally a great idea.
+ * in a failure state isn't generally a great idea and if it fails, we want a
+ * fall back position using caller supplied buffers.
+ *
+ * If the return value is different from the buffer supplied, it needs to be
+ * freed by the caller.
  */
-static const char *print_mem_maybe_null(const void *s, size_t n,
-                                        char out[MEM_BUFFER_SIZE])
+static char *print_mem_maybe_null(const void *s, size_t n,
+                                        char outbuf[MEM_BUFFER_SIZE])
 {
     size_t i;
     const unsigned char *p = (const unsigned char *)s;
-    int pad = 2*n >= MEM_BUFFER_SIZE;
+    char *out = outbuf;
+    int pad = 2 * n >= MEM_BUFFER_SIZE;
 
     if (s == NULL)
-        return "(NULL)";
-    if (pad)
-        n = MEM_BUFFER_SIZE-4;
+        return memcpy(outbuf, "(NULL)", sizeof("(NULL)") - 1);
+    if (pad) {
+        if ((out = OPENSSL_malloc(2 * n + 1)) == NULL) {
+            out = outbuf;
+            n = (MEM_BUFFER_SIZE - 4) / 2;
```

Comment:
Three dots and a terminator is four characters.  Lines 396 - 401.
It would be possible to round better here to gain a single character if the buffer length is odd.

Still, this should be a rare case -- the buffer to be displayed has to be larger than the constant size and malloc has to fail.  There isn't much else to do except truncate.